File: 890.Find-and-Replace-Pattern/findAndPeplacePattern.py
"""
本质上是一种对应关系；可以对应到字母上，也可以对应到数字上。

之前的错误在于：
当两个连续的对应关系相同时，我就让程序退出了，忽略了之后的序列---导致错误，没有遍历完成
"""
class Solution:
    def findAndReplacePattern(self, words, pattern):
        """
        :type words: List[str]
        :type pattern: str
        :rtype: List[str]
        """
        p = self.getPattern(pattern)
        res = []
        
        for word in words:
            if self.getPattern(word) == p:
                res.append(word)
        
        return res
        
    def getPattern(self, pattern):
        patternHash = {}
        p = ""
        idx = 0
        for i in pattern:
            if i not in patternHash:
                patternHash[i] = idx
                idx += 1
            p += str(patternHash[i])
        return p
            

File: 890.Find-and-Replace-Pattern/findAndPeplacePattern.cpp
/**
 * unordered_map<char,string> 字典
 * 插入时：
 * unordered_map.insert({'a',"b"});字典里面是逗号分割
 */
class Solution {
public:
    vector<string> findAndReplacePattern(vector<string>& words, string b) {
        vector<string> res;
        string p = isSame(b);
        for(auto ele : words){
            if(isSame(ele) == p)
                res.push_back(ele);
        }
        return res;
    }
    string isSame(string a){
        unordered_map<char,string> to;
        string ret = "";
        int idx = 0;
        for(auto w:a){
            if(to.find(w) == to.end())
                to.insert({w,to_string(idx++)});
            ret += to.find(w)->second;
        }
        return ret;
    }
};

